internal/controller/pkg/revision/reconciler.go

```diff
@@ -623,3 +634,21 @@ func (r *Reconciler) Reconcile(ctx context.Context, req reconcile.Request) (reco
 	pr.SetConditions(v1.Healthy())
 	return reconcile.Result{Requeue: false}, errors.Wrap(r.client.Status().Update(ctx, pr), errUpdateStatus)
 }
+
+func (r *Reconciler) deactivateRevision(ctx context.Context, pr v1.PackageRevision) error {
+	// If we are inactive, we want to remove self from the lock.
+	// If we skipped dependency resolution, we will not be present in the lock.
+	if err := r.lock.RemoveSelf(ctx, pr); err != nil {
+		return errors.Wrap(err, errRemoveLock)
+	}
+
+	if err := r.objects.Relinquish(ctx, pr); err != nil {
+		return errors.Wrap(err, "cannot relinquish objects")
+	}
+
+	if err := r.hook.Deactivate(ctx, pr); err != nil {
+		return errors.Wrap(err, "deactivation hook failed")
+	}
```

Comment:
@turkenh these steps are already carried out by the existing establisher and hook logic when a revision is inactive. Can we get the same outcome here by just adding a check [here](https://github.com/crossplane/crossplane/blob/e40d04c66f0783352d6970809f30d75e3d396d8d/internal/controller/pkg/revision/reconciler.go#L559) and skip dependency resolution if the package is inactive?

Reply:
@hasheddan just to be clear, I am not trying to solve the specific problem in #3742; instead trying to fix the general class of issues where an inactive provider revision reconciliation might fail early in the reconciliation and not be deactivated successfully, including ensuring deactivation is completed before deletion, i.e. finalizer removed.

When I checked the `ProviderRevision` reconciler, there is [a bunch of code](https://github.com/crossplane/crossplane/blob/e40d04c66f0783352d6970809f30d75e3d396d8d/internal/controller/pkg/revision/reconciler.go#L407-L584) that needs to be executed successfully to reach until the establish step (as `inactive`). Apart from the dependency-related steps, e.g. [xp constraint check](https://github.com/crossplane/crossplane/blob/e40d04c66f0783352d6970809f30d75e3d396d8d/internal/controller/pkg/revision/reconciler.go#L542) and [resolution](https://github.com/crossplane/crossplane/blob/e40d04c66f0783352d6970809f30d75e3d396d8d/internal/controller/pkg/revision/reconciler.go#L559), an inactive revision might also fail in other steps, e.g. while [fetching](https://github.com/crossplane/crossplane/blob/e40d04c66f0783352d6970809f30d75e3d396d8d/internal/controller/pkg/revision/reconciler.go#L419) or [parsing the package](https://github.com/crossplane/crossplane/blob/e40d04c66f0783352d6970809f30d75e3d396d8d/internal/controller/pkg/revision/reconciler.go#LL478C26-L478C26), imagine an old revision package is removed from the registry and no longer in the cache. Currently, we still have to successfully pull an inactive revision just to leave control of resources. 

Imagine the following hypothetical scenario:

1. User installs provider-foo v1.
2. v1 version got deleted from the registry
3. User updates provider-foo to an existing version, say v2.

IIUC, if package v1 is not in the cache somehow, e.g. XP restarted, the user will not be able to migrate to v2 since the reconciliation of v1 inactive will fail early and they will get the "is already controlled by ProviderRevision" error. 
This is just one example of how the early steps might fail for an inactive revision.

In this PR, I am trying to leverage the fact that our controller has reference to all objects it owns/controls. So, we don't need to fetch and parse the whole package just to remove the `controller: false` from those objects.

Reply:
> When I checked the ProviderRevision reconciler, there is [a bunch of code](https://github.com/crossplane/crossplane/blob/e40d04c66f0783352d6970809f30d75e3d396d8d/internal/controller/pkg/revision/reconciler.go#L407-L584) that needs to be executed successfully to reach until the establish step (as inactive). Apart from the dependency-related steps, e.g. [xp constraint check](https://github.com/crossplane/crossplane/blob/e40d04c66f0783352d6970809f30d75e3d396d8d/internal/controller/pkg/revision/reconciler.go#L542) and [resolution](https://github.com/crossplane/crossplane/blob/e40d04c66f0783352d6970809f30d75e3d396d8d/internal/controller/pkg/revision/reconciler.go#L559), an inactive revision might also fail in other steps, e.g. while [fetching](https://github.com/crossplane/crossplane/blob/e40d04c66f0783352d6970809f30d75e3d396d8d/internal/controller/pkg/revision/reconciler.go#L419) or [parsing the package](https://github.com/crossplane/crossplane/blob/e40d04c66f0783352d6970809f30d75e3d396d8d/internal/controller/pkg/revision/reconciler.go#LL478C26-L478C26), imagine an old revision package is removed from the registry and no longer in the cache. Currently, we still have to successfully pull an inactive revision just to leave control of resources.

It's worth noting that a large portion of this code is not executed on every reconciliation -- the vast majority of cases all of this translates to just reading a file from disk and parsing, which has a very low failure rate (note: I still would like to optimize this in the future. If we parse successfully and save to disk then we really don't need to be parsing every time. I have thought about just serializing Go structs for the cache, but I think that it outside the scope of this specific PR). If you also skip dependency resolution for an inactive revision as I recommended above, then reading and parsing is really the only operation in this entire block.

If a package no longer exists in the cache or in the registry, which I would also view as a fairly infrequent case, then we have a bigger problem because the actual content is unknown. We cannot count on the status to always reflect the proper state as it could have been mutated or lost. In this event, I think we have two options (not mutually exclusive):
- Add tooling to help users clean up dangling controller references and manually remove finalizers.
- Allow for deletion even if the ownership of objects is unable to be transitioned (which I suggest adding in https://github.com/crossplane/crossplane/issues/3742), then update the dependency manager (or add another component) that watches for failing revisions and triggers a manual cleanup of controller references if the controlling revision no longer exists.

Reply:
Just prepared the following to visualize what I am proposing here:

![Package Revision Reconciler](https://github.com/crossplane/crossplane/assets/9900707/65cde448-23f3-4b72-a83f-013c41cb0089)

Reply:
Marking this thread as resolved, feel free to unresolve it or start another one.